src/mlpack/core/tree/binary_space_tree/binary_space_tree.hpp

```diff
@@ -333,6 +333,36 @@ class BinarySpaceTree
   size_t NumChildren() const;
 
   /**
+   * Return the nearest child node to the given query point.  If this is a leaf
+   * node, it will return a reference to itself.
+   */
+  template<typename VecType>
+  BinarySpaceTree& GetNearestChild(
+      const VecType& point,
+      typename boost::enable_if<IsVector<VecType> >::type* = 0);
+
+  /**
+   * Return the furthest child node to the given query point.  If this is a leaf
+   * node, it will return a reference to itself.
+   */
+  template<typename VecType>
+  BinarySpaceTree& GetFurthestChild(
+      const VecType& point,
+      typename boost::enable_if<IsVector<VecType> >::type* = 0);
+
+  /**
+   * Return the nearest child node to the given query node.  If it can't decide
+   * it will return a null pointer.
+   */
+  BinarySpaceTree* GetNearestChild(const BinarySpaceTree& queryNode);
+
+  /**
+   * Return the furthest child node to the given query node.  If it can't decide
+   * it will return a null pointer.
+   */
+  BinarySpaceTree* GetFurthestChild(const BinarySpaceTree& queryNode);
```

Comment:
I don't think you're using these overloads anywhere. Can I suggest returning a `size_t` instead of a reference or pointer? You can return `NumChildren()` to indicate an invalid child. This avoids the issues with returning `*this` and there's never a need to return `NULL`. The user can check on their own whether or not the results are valid and we also manage to keep our TreeType API simpler. What do you think?